bundles/org.openhab.binding.hapero/src/main/java/org/openhab/binding/hapero/internal/handler/HaperoBridgeHandler.java

```diff
@@ -0,0 +1,615 @@
+        }
+
+        /*
+         * If the update was successful and the bridge was offline
+         * because of some earlier communication error, set it back to online now.
+         */
+        if (getThing().getStatus() != ThingStatus.ONLINE && !faultOcurred) {
+            updateStatus(ThingStatus.ONLINE);
+        }
+    }
+
+    /**
+     * Gets the device data from either the FTP Server or the file system
+     * and calls {@link updateDeviceList}. Afterwards, triggers an channel data update for all Things.
+     */
+    private void updateThings() {
+        HaperoConfiguration config = getConfigAs(HaperoConfiguration.class);
+
+        /* Get Data from local filesystem */
+        if (config.accessMode.contentEquals(HaperoBindingConstants.CONFIG_ACCESS_FILESYSTEM)) {
+            Path filePath = null;
+            try {
+                filePath = Path.of(config.fileStoragePath, HaperoBindingConstants.DATA_FILENAME);
+            } catch (Exception e) {
+                logger.debug("Exception in updateThings: {}", e.getMessage());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "@text/offline.file.read.error");
+                return;
+            }
+
+            updateDeviceList(filePath);
+        } else if (config.accessMode.contentEquals(HaperoBindingConstants.CONFIG_ACCESS_FTP)) {
+            /* Get data from FTP */
+            InputStream stream;
+            if (polledDirectory != null) {
+                try {
+                    stream = polledDirectory.getFileStream(HaperoBindingConstants.DATA_FILENAME);
+                } catch (IOException e) {
+                    logger.debug("Exception in updateThings: {}", e.getMessage());
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "@text/offline.ftp.read.error");
+                    return;
+                }
+
+                updateDeviceList(stream);
+            } else {
+                logger.debug("PolledDirectory null in updateThings");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "@text/offline.ftpmonitoring.failed");
+            }
+        }
+
+        /* Notify all things to update their channels */
+        for (HaperoThingHandler handler : devicesToNotify) {
+            handler.updateAllChannels();
+            handler.updateThingProperties();
+        }
+    }
+
+    /**
+     * Starts a Runnable with delay time refreshTimeout.
+     * If the Runnable is triggered, set the Bridge status to offline.
+     * The Runnable is restarted every time a FTP File update is detected.
+     */
+    private void startFtpTimeout() {
+        HaperoConfiguration config = getConfigAs(HaperoConfiguration.class);
+
+        if (null != ftpTimeoutJob) {
+            ftpTimeoutJob.cancel(true);
+        }
+
+        ftpTimeoutJob = scheduler.schedule(() -> {
```

Comment:
Actually, i am getting a couple of null annotation warnings that i cant for the life of me figure out how to get rid of:

'''
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[162,27] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[187,29] Potential null pointer access: this expression has type 'capture#2-of ?', a free type variable that may represent a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[211,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[355,25] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[547,30] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[579,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[592,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[597,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[602,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] C:\openhab-dev\openhab-main\git\openhab-addons\bundles\org.openhab.binding.hapero\src\main\java\org\openhab\binding\hapero\internal\handler\HaperoBridgeHandler.java:[608,17] Potential null pointer access: this expression has a '@Nullable' type
'''
I am not a professional Java developer (i develop in C# and C at work) and googeling did not really help me because it seems to be something specific with those types, at least i could not find the right words to search for. So if you could provide me a hint, i would appreciate it!

Reply:
I'm also professionally a C# developer. So like me, you'll be able to preview features in C# years before they arrive in Java, and you'll be ready for them. 😄

Null annotations were a little hard also for me to get used to though. In this case it's because you are referencing a `@Nullable` object without checking first if it's `null`. But wait, you did check that, you say? Yes, but after you checked it, it could be `null` again - because some other thread changed it. So for instance variables you need to assign to a local variable in order to guarantee that your check will ensure you are working with a non-null reference.